src/profiler.cpp

```diff
@@ -1647,6 +1652,108 @@ void Profiler::dumpText(Writer& out, Arguments& args) {
     }
 }
 
+static void recordSampleType(ProtoBuffer& otlp_buffer, Engine* engine, Index& strings, const char* unit) {
+    using namespace Otlp;
+    protobuf_mark_t sample_type_mark = otlp_buffer.startMessage(Profile::sample_type);
+    otlp_buffer.field(ValueType::type_strindex, strings.indexOf(engine->type()));
+    otlp_buffer.field(ValueType::unit_strindex, strings.indexOf(unit));
+    otlp_buffer.field(ValueType::aggregation_temporality, AggregationTemporality::cumulative);
+    otlp_buffer.commitMessage(sample_type_mark);
+}
+
+void Profiler::dumpOtlp(Writer& out, Arguments& args) {
+    using namespace Otlp;
+    ProtoBuffer otlp_buffer{OTLP_BUFFER_INITIAL_SIZE};
+    Index strings;
+    Index functions;
+
+    protobuf_mark_t resource_profiles_mark = otlp_buffer.startMessage(ProfilesData::resource_profiles);
+    protobuf_mark_t scope_profiles_mark = otlp_buffer.startMessage(ResourceProfiles::scope_profiles);
+    protobuf_mark_t profile_mark = otlp_buffer.startMessage(ScopeProfiles::profiles);
+
+    recordSampleType(otlp_buffer, _engine, strings, "count");
+    recordSampleType(otlp_buffer, _engine, strings, _engine->units());
+
+    std::vector<CallTraceSample*> callTraceSamples;
+    _call_trace_storage.collectSamples(callTraceSamples);
+
+    struct SampleInfo {
+        u64 counter;
+        u64 samples;
+        u32 num_frames;
+    };
+    std::vector<SampleInfo> samplesInfo;
+    samplesInfo.reserve(callTraceSamples.size());
+
+    FrameName fn(args, args._style & ~STYLE_ANNOTATE, _epoch, _thread_names_lock, _thread_names);
+    protobuf_mark_t location_indices_mark = otlp_buffer.startMessage(Profile::location_indices);
+    for (const auto& callTraceSample : callTraceSamples) {
+        CallTrace* trace = callTraceSample->acquireTrace();
+        if (trace == NULL || excludeTrace(&fn, trace) || callTraceSample->samples == 0) continue;
+
+        samplesInfo.push_back(SampleInfo{callTraceSample->samples, callTraceSample->counter, (u32) trace->num_frames});
+        for (u32 j = 0; j < trace->num_frames; ++j) {
+            u32 function_idx = functions.indexOf(fn.name(trace->frames[j]));
+            otlp_buffer.putVarInt(function_idx);
+        }
+    }
+    otlp_buffer.commitMessage(location_indices_mark);
+
+    u64 frames_seen = 0;
+    for (const SampleInfo& sampleInfo : samplesInfo) {
+        protobuf_mark_t sample_mark = otlp_buffer.startMessage(Profile::sample);
+        otlp_buffer.field(Sample::locations_start_index, frames_seen);
+        otlp_buffer.field(Sample::locations_length, sampleInfo.num_frames);
+
+        protobuf_mark_t sample_value_mark = otlp_buffer.startMessage(Sample::value);
```

Comment:
There are many tiny messages whose length can be encoded in 1 byte.
Do you want to provide a version of `startMessage` (or add an optional argument) to specify maximum length?